File: i2pchat/storage/profile_dat.py
"""
Encrypted at-rest storage for the working profile identity ``.dat`` file.

Format (binary):
  4 bytes  — magic ``I2PK``
  2 bytes  — format version (big-endian uint16, currently 1)
  32 bytes — random salt
  rest     — NaCl SecretBox ciphertext wrapping UTF-8 private-key line
             (optional trailing ``\\n`` + legacy peer line is accepted on decrypt
             for migration, but new writes store only the key line)

Wrapping key:
  - Preferred: OS keyring entry ``i2pchat`` / ``{profile}__dat_wrap__``
    (base64 of 32 random bytes).
  - Fallback when keyring is unavailable: ``{profile}.dat.wrap`` next to the
    ``.dat`` (mode 0600). Same local threat model as former plaintext ``.dat``,
    but a leaked ``.dat`` alone (without the wrap sidecar / keyring) is useless.

Legacy plaintext UTF-8 ``.dat`` files are still readable; callers should
re-write them encrypted after a successful load (transparent migration).
"""

from __future__ import annotations

import base64
import logging
import os
import secrets
import struct
from typing import Callable, Optional

from i2pchat import crypto
from i2pchat.storage.blindbox_state import atomic_write_bytes, atomic_write_text

logger = logging.getLogger("i2pchat.storage.profile_dat")

PROFILE_DAT_MAGIC = b"I2PK"
PROFILE_DAT_VERSION = 1
PROFILE_DAT_SALT_SIZE = 32
PROFILE_DAT_HEADER_SIZE = 4 + 2 + PROFILE_DAT_SALT_SIZE

DAT_WRAP_KEYRING_SUFFIX = "__dat_wrap__"
KEYRING_SERVICE = "i2pchat"


def is_encrypted_profile_dat(raw: bytes) -> bool:
    return len(raw) >= PROFILE_DAT_HEADER_SIZE and raw[:4] == PROFILE_DAT_MAGIC


def profile_dat_wrap_path(profile_data_dir: str, profile: str) -> str:
    return os.path.join(profile_data_dir, f"{profile}.dat.wrap")


def _try_keyring_get_wrap(profile: str) -> Optional[bytes]:
    try:
        import keyring

        raw = keyring.get_password(KEYRING_SERVICE, f"{profile}{DAT_WRAP_KEYRING_SUFFIX}")
    except ImportError:
        return None
    except Exception as e:
        logger.debug("keyring wrap get failed (%s): %s", profile, e)
        return None
    if not raw:
        return None
    try:
        key = base64.b64decode(raw.encode("ascii"), validate=True)
    except Exception:
        return None
    if len(key) != 32:
        return None
    return key


def _try_keyring_set_wrap(profile: str, wrap_key: bytes) -> bool:
    if len(wrap_key) != 32:
        raise ValueError("wrap_key must be 32 bytes")
    try:
        import keyring

        keyring.set_password(
            KEYRING_SERVICE,
            f"{profile}{DAT_WRAP_KEYRING_SUFFIX}",
            base64.b64encode(wrap_key).decode("ascii"),
        )
        return True
    except ImportError:
        return False
    except Exception as e:
        logger.debug("keyring wrap set failed (%s): %s", profile, e)
        return False


def _read_wrap_sidecar(path: str) -> Optional[bytes]:
    if not os.path.isfile(path):
        return None
    try:
        with open(path, "r", encoding="utf-8") as f:
            token = f.read().strip()
        key = base64.b64decode(token.encode("ascii"), validate=True)
    except Exception:
        return None
    if len(key) != 32:
        return None
    return key


def _write_wrap_sidecar(path: str, wrap_key: bytes) -> None:
    atomic_write_text(path, base64.b64encode(wrap_key).decode("ascii") + "\n")


def get_or_create_dat_wrap_key(profile: str, profile_data_dir: str) -> bytes:
    """Return a stable 32-byte wrap key for this profile (keyring or sidecar)."""
    existing = _try_keyring_get_wrap(profile)
    if existing is not None:
        return existing
    sidecar = profile_dat_wrap_path(profile_data_dir, profile)
    existing = _read_wrap_sidecar(sidecar)
    if existing is not None:
        # Best-effort promote sidecar → keyring for stronger protection.
        _try_keyring_set_wrap(profile, existing)
        return existing
    wrap_key = secrets.token_bytes(32)
    stored_in_keyring = _try_keyring_set_wrap(profile, wrap_key)
    _write_wrap_sidecar(sidecar, wrap_key)
    if stored_in_keyring:
        logger.info("Profile .dat wrap key stored in OS keyring (+ local sidecar)")
    else:
        logger.info(
            "Profile .dat wrap key stored in local sidecar (OS keyring unavailable)"
        )
    return wrap_key


def load_dat_wrap_key(profile: str, profile_data_dir: str) -> Optional[bytes]:
    """Load an existing wrap key without creating one."""
    existing = _try_keyring_get_wrap(profile)
    if existing is not None:
        return existing
    return _read_wrap_sidecar(profile_dat_wrap_path(profile_data_dir, profile))


def _derive_file_key(wrap_key: bytes, salt: bytes) -> bytes:
    prk = crypto.hkdf_extract(b"I2PCHAT-PROFILE-DAT", wrap_key)
    profile_key = crypto.hkdf_expand(prk, b"I2PCHAT-PROFILE-DAT|profile-key", 32)
    prk2 = crypto.hkdf_extract(salt, profile_key)
    return crypto.hkdf_expand(prk2, b"I2PCHAT-PROFILE-DAT|file-key", 32)


def encrypt_profile_dat(private_key_base64: str, wrap_key: bytes) -> bytes:
    if not crypto.NACL_AVAILABLE:
        raise RuntimeError("PyNaCl is required to encrypt profile .dat")
    key = (private_key_base64 or "").strip()
    if not key:
        raise ValueError("private_key_base64 is empty")
    salt = secrets.token_bytes(PROFILE_DAT_SALT_SIZE)
    file_key = _derive_file_key(wrap_key, salt)
    plaintext = (key + "\n").encode("utf-8")
    ciphertext = crypto.encrypt_message(file_key, plaintext)
    return PROFILE_DAT_MAGIC + struct.pack(">H", PROFILE_DAT_VERSION) + salt + ciphertext


def decrypt_profile_dat(raw: bytes, wrap_key: bytes) -> str:
    if not is_encrypted_profile_dat(raw):
        raise ValueError("Not an encrypted profile .dat")
    if not crypto.NACL_AVAILABLE:
        raise RuntimeError("PyNaCl is required to decrypt profile .dat")
    version = struct.unpack(">H", raw[4:6])[0]
    if version != PROFILE_DAT_VERSION:
        raise ValueError(f"Unsupported encrypted profile .dat version {version}")
    salt = raw[6:PROFILE_DAT_HEADER_SIZE]
    ciphertext = raw[PROFILE_DAT_HEADER_SIZE:]
    file_key = _derive_file_key(wrap_key, salt)
    plaintext = crypto.decrypt_message(file_key, ciphertext)
    if plaintext is None:
        raise ValueError("Profile .dat decryption failed (wrong wrap key or tampered)")
    text = plaintext.decode("utf-8")
    lines = [ln.strip() for ln in text.splitlines() if ln.strip()]
    if not lines:
        raise ValueError("Decrypted profile .dat is empty")
    return lines[0]


def parse_plaintext_profile_dat(
    text: str,
    *,
    is_probable_peer: Optional[Callable[[str], bool]] = None,
) -> tuple[Optional[str], Optional[str]]:
    """Parse legacy UTF-8 ``.dat`` into (private_key_b64 | None, legacy_peer | None)."""
    lines = [ln.strip() for ln in (text or "").splitlines() if ln.strip()]
    if not lines:
        return None, None
    peer_check = is_probable_peer or (lambda _s: False)
    private_key: Optional[str] = None
    legacy_peer: Optional[str] = None
    if not peer_check(lines[0]):
        private_key = lines[0]
        if len(lines) > 1 and peer_check(lines[1]):
            legacy_peer = lines[1]
    elif peer_check(lines[0]):
        # keyring-only scenario: file held only the locked peer
        legacy_peer = lines[0]
    return private_key, legacy_peer


def read_profile_dat_file(
    path: str,
    *,
    profile: str,
    profile_data_dir: str,
    is_probable_peer: Optional[Callable[[str], bool]] = None,
    create_wrap_key: bool = True,
) -> tuple[Optional[str], Optional[str], bool]:
    """
    Read a profile ``.dat``.

    Returns ``(private_key_b64 | None, legacy_peer | None, was_plaintext)``.
    ``was_plaintext`` is True when the on-disk file was legacy UTF-8 (caller
    should re-encrypt).
    """
    if not os.path.isfile(path):
        return None, None, False
    with open(path, "rb") as f:
        raw = f.read()
    if not raw:
        return None, None, False
    if is_encrypted_profile_dat(raw):
        wrap = (
            get_or_create_dat_wrap_key(profile, profile_data_dir)
            if create_wrap_key
            else load_dat_wrap_key(profile, profile_data_dir)
        )
        if wrap is None:
            raise ValueError(
                f"Encrypted profile .dat at {path} but no wrap key is available "
                "(OS keyring entry missing and no .dat.wrap sidecar)"
            )
        key = decrypt_profile_dat(raw, wrap)
        return key, None, False
    # Legacy plaintext UTF-8
    try:
        text = raw.decode("utf-8")
    except UnicodeDecodeError as e:
        raise ValueError(f"Profile .dat is neither encrypted nor valid UTF-8: {e}") from e
    private_key, legacy_peer = parse_plaintext_profile_dat(
        text, is_probable_peer=is_probable_peer
    )
    return private_key, legacy_peer, True


def write_encrypted_profile_dat(
    path: str,
    private_key_base64: str,
    *,
    profile: str,
    profile_data_dir: str,
) -> None:
    """Encrypt and atomically write the identity key; creates wrap key if needed."""
    if not crypto.NACL_AVAILABLE:
        # Last-resort fallback: keep previous plaintext behaviour so the app
        # remains usable without PyNaCl (should be rare in real installs).
        logger.warning(
            "PyNaCl unavailable — writing profile .dat without at-rest encryption"
        )
        atomic_write_text(path, (private_key_base64 or "").strip() + "\n")
        return
    wrap = get_or_create_dat_wrap_key(profile, profile_data_dir)
    # Always keep a 0600 sidecar so a copied profile directory remains openable
    # even when OS keyring is unavailable on the destination machine. A leaked
    # .dat alone (without .dat.wrap / keyring) stays opaque.
    sidecar = profile_dat_wrap_path(profile_data_dir, profile)
    if _read_wrap_sidecar(sidecar) != wrap:
        _write_wrap_sidecar(sidecar, wrap)
    blob = encrypt_profile_dat(private_key_base64, wrap)
    atomic_write_bytes(path, blob)


def plaintext_key_bytes_for_backup(
    path: str,
    *,
    profile: str,
    profile_data_dir: str,
) -> bytes:
    """
    Return UTF-8 plaintext key bytes suitable for inclusion in a passphrase-
    protected backup bundle (portable across machines without the wrap key).
    """
    key, _peer, _was_plain = read_profile_dat_file(
        path,
        profile=profile,
        profile_data_dir=profile_data_dir,
        create_wrap_key=False,
    )
    if not key:
        raise ValueError(f"No private key found in {path}")
    return (key.strip() + "\n").encode("utf-8")

File: cpp/core/src/storage/profile_dat.cpp
#include "i2pchat/storage/profile_dat.hpp"

#include <algorithm>
#include <cctype>
#include <sstream>
#include <vector>

#include "i2pchat/crypto.hpp"
#include "i2pchat/encoding.hpp"
#include "i2pchat/storage/atomic_write.hpp"
#include "i2pchat/storage/keyring.hpp"

namespace i2pchat::storage {
namespace {

std::string trim(std::string_view text) {
    std::size_t begin = 0;
    while (begin < text.size() &&
           std::isspace(static_cast<unsigned char>(text[begin])) != 0) {
        ++begin;
    }
    std::size_t end = text.size();
    while (end > begin && std::isspace(static_cast<unsigned char>(text[end - 1])) != 0) {
        --end;
    }
    return std::string(text.substr(begin, end - begin));
}

std::vector<std::string> non_empty_lines(std::string_view text) {
    std::vector<std::string> lines;
    std::istringstream stream{std::string(text)};
    std::string line;
    while (std::getline(stream, line)) {
        const std::string trimmed = trim(line);
        if (!trimmed.empty()) {
            lines.push_back(trimmed);
        }
    }
    return lines;
}

/// The wrap key travels as standard base64 of 32 raw bytes, in both the keyring
/// and the sidecar, because that is what Python's implementation writes.
std::optional<Bytes> decode_wrap_key(std::string_view token) {
    const std::optional<Bytes> decoded = encoding::base64_decode(trim(token));
    if (!decoded.has_value() || decoded->size() != 32) {
        return std::nullopt;
    }
    return decoded;
}

std::optional<Bytes> keyring_wrap_key(std::string_view profile) {
    const std::optional<std::string> stored =
        keyring::get(kKeyringService, dat_wrap_keyring_account(profile));
    if (!stored.has_value()) {
        return std::nullopt;
    }
    return decode_wrap_key(*stored);
}

bool store_keyring_wrap_key(std::string_view profile, ByteView wrap_key) {
    return keyring::set(kKeyringService, dat_wrap_keyring_account(profile),
                        encoding::base64_encode(wrap_key));
}

std::optional<Bytes> read_wrap_sidecar(const std::filesystem::path& path) {
    if (!std::filesystem::is_regular_file(path)) {
        return std::nullopt;
    }
    try {
        return decode_wrap_key(to_string(ByteView(read_file(path))));
    } catch (const std::exception&) {
        return std::nullopt;
    }
}

void write_wrap_sidecar(const std::filesystem::path& path, ByteView wrap_key) {
    atomic_write_text(path, encoding::base64_encode(wrap_key) + "\n");
}

}  // namespace

bool is_encrypted_profile_dat(ByteView raw) {
    return raw.size() >= kProfileDatHeaderSize &&
           std::equal(kProfileDatMagic.begin(), kProfileDatMagic.end(), raw.begin());
}

std::filesystem::path profile_dat_wrap_path(const std::filesystem::path& profile_data_dir,
                                            std::string_view profile) {
    return profile_data_dir / (std::string(profile) + ".dat.wrap");
}

std::string dat_wrap_keyring_account(std::string_view profile) {
    return std::string(profile) + std::string(kDatWrapKeyringSuffix);
}

Bytes get_or_create_dat_wrap_key(std::string_view profile,
                                 const std::filesystem::path& profile_data_dir) {
    if (const std::optional<Bytes> existing = keyring_wrap_key(profile)) {
        return *existing;
    }

    const std::filesystem::path sidecar = profile_dat_wrap_path(profile_data_dir, profile);
    if (const std::optional<Bytes> existing = read_wrap_sidecar(sidecar)) {
        // Promote the sidecar into the keyring, where the OS can protect it.
        (void)store_keyring_wrap_key(profile, ByteView(*existing));
        return *existing;
    }

    const Bytes wrap_key = crypto::random_bytes(32);
    (void)store_keyring_wrap_key(profile, ByteView(wrap_key));
    write_wrap_sidecar(sidecar, ByteView(wrap_key));
    return wrap_key;
}

std::optional<Bytes> load_dat_wrap_key(std::string_view profile,
                                       const std::filesystem::path& profile_data_dir) {
    if (const std::optional<Bytes> existing = keyring_wrap_key(profile)) {
        return existing;
    }
    return read_wrap_sidecar(profile_dat_wrap_path(profile_data_dir, profile));
}

Bytes derive_profile_dat_file_key(ByteView wrap_key, ByteView salt) {
    const Bytes prk = crypto::hkdf_extract(as_bytes(kProfileDatDomain), wrap_key);
    const Bytes profile_key = crypto::hkdf_expand(
        ByteView(prk), as_bytes(std::string(kProfileDatDomain) + "|profile-key"), 32);
    const Bytes prk2 = crypto::hkdf_extract(salt, ByteView(profile_key));
    return crypto::hkdf_expand(
        ByteView(prk2), as_bytes(std::string(kProfileDatDomain) + "|file-key"), 32);
}

Bytes encrypt_profile_dat(std::string_view private_key_base64, ByteView wrap_key) {
    const std::string key = trim(private_key_base64);
    if (key.empty()) {
        throw ProfileDatError("private_key_base64 is empty");
    }
    const Bytes salt = crypto::random_bytes(kProfileDatSaltSize);
    const Bytes file_key = derive_profile_dat_file_key(wrap_key, ByteView(salt));
    const Bytes ciphertext =
        crypto::encrypt_message(ByteView(file_key), as_bytes(key + "\n"));

    Bytes blob;
    blob.reserve(kProfileDatHeaderSize + ciphertext.size());
    append(blob, kProfileDatMagic);
    append_u16_be(blob, kProfileDatVersion);
    append(blob, ByteView(salt));
    append(blob, ByteView(ciphertext));
    return blob;
}

std::string decrypt_profile_dat(ByteView raw, ByteView wrap_key) {
    if (!is_encrypted_profile_dat(raw)) {
        throw ProfileDatError("Not an encrypted profile .dat");
    }
    const std::uint16_t version = read_u16_be(raw.subspan(4, 2));
    if (version != kProfileDatVersion) {
        throw ProfileDatError("Unsupported encrypted profile .dat version " +
                              std::to_string(version));
    }

    const ByteView salt = raw.subspan(6, kProfileDatSaltSize);
    const ByteView ciphertext = raw.subspan(kProfileDatHeaderSize);
    const Bytes file_key = derive_profile_dat_file_key(wrap_key, salt);

    const std::optional<Bytes> plaintext =
        crypto::decrypt_message(ByteView(file_key), ciphertext);
    if (!plaintext.has_value()) {
        throw ProfileDatError(
            "Profile .dat decryption failed (wrong wrap key or tampered)");
    }

    const std::vector<std::string> lines =
        non_empty_lines(to_string(ByteView(*plaintext)));
    if (lines.empty()) {
        throw ProfileDatError("Decrypted profile .dat is empty");
    }
    return lines.front();
}

ProfileDatContents parse_plaintext_profile_dat(std::string_view text,
                                               const PeerAddressPredicate& is_peer) {
    ProfileDatContents contents;
    const std::vector<std::string> lines = non_empty_lines(text);
    if (lines.empty()) {
        return contents;
    }

    const auto peer_check = [&is_peer](const std::string& line) {
        return is_peer ? is_peer(line) : false;
    };

    if (!peer_check(lines[0])) {
        contents.private_key_base64 = lines[0];
        if (lines.size() > 1 && peer_check(lines[1])) {
            contents.legacy_peer = lines[1];
        }
    } else {
        // A keyring-only profile: the file held nothing but the locked peer.
        contents.legacy_peer = lines[0];
    }
    return contents;
}

ProfileDatContents read_profile_dat_file(const std::filesystem::path& path,
                                         std::string_view profile,
                                         const std::filesystem::path& profile_data_dir,
                                         const PeerAddressPredicate& is_peer,
                                         bool create_wrap_key) {
    ProfileDatContents contents;
    if (!std::filesystem::is_regular_file(path)) {
        return contents;
    }
    const Bytes raw = read_file(path);
    if (raw.empty()) {
        return contents;
    }

    if (is_encrypted_profile_dat(ByteView(raw))) {
        const std::optional<Bytes> wrap =
            create_wrap_key
                ? std::optional<Bytes>(get_or_create_dat_wrap_key(profile, profile_data_dir))
                : load_dat_wrap_key(profile, profile_data_dir);
        if (!wrap.has_value()) {
            throw ProfileDatError(
                "Encrypted profile .dat at " + path.string() +
                " but no wrap key is available (no keyring entry and no .dat.wrap)");
        }
        contents.private_key_base64 =
            decrypt_profile_dat(ByteView(raw), ByteView(*wrap));
        return contents;
    }

    contents = parse_plaintext_profile_dat(to_string(ByteView(raw)), is_peer);
    contents.was_plaintext = true;
    return contents;
}

void write_encrypted_profile_dat(const std::filesystem::path& path,
                                 std::string_view private_key_base64,
                                 std::string_view profile,
                                 const std::filesystem::path& profile_data_dir) {
    const Bytes wrap = get_or_create_dat_wrap_key(profile, profile_data_dir);

    // Keep the sidecar in step with the key actually in use, so a copied profile
    // directory stays openable on a machine with no keyring.
    const std::filesystem::path sidecar = profile_dat_wrap_path(profile_data_dir, profile);
    if (read_wrap_sidecar(sidecar) != std::optional<Bytes>(wrap)) {
        write_wrap_sidecar(sidecar, ByteView(wrap));
    }

    atomic_write_bytes(path,
                       ByteView(encrypt_profile_dat(private_key_base64, ByteView(wrap))));
}

}  // namespace i2pchat::storage
